File: analysis/ana01.R
require(rio)
require(dplyr)
require(tidyr)
require(slider)
require(purrr)
require(stringr)
require(ggplot2)


perform_summ <- function(d, high_pattern){

  
  
  d2 <- d %>%
    select(which(!apply(is.na(d), 2, all))) %>%
    mutate(
      triplet = paste(
        lag(correct_answer_index, 2), lag(correct_answer_index), correct_answer_index, sep = ""
      )
      
    ) %>% 
    select(orientation_index,
           correct_answer_index,
           triplet,
           key_resp.corr,
           key_resp.rt,
           thisTrialN)%>%
    mutate(type = if_else(thisTrialN %% 2 == 1, 'Pat','Ran'),
           block = 1 + (row_number() - 1) %/% 80, 
           epoch = 1 + (row_number() - 1) %/% 400) %>%
    mutate( frequency = if_else(str_detect(triplet, high_pattern), 'High', 'Low')) %>%
    filter( key_resp.corr == 1)
  
  
  
  
  # Assuming your dataframe is named 'd.learn'
  d2.counts <- d2 %>%
    filter(!str_detect(triplet,'NA')) %>%
    group_by(epoch, frequency,  
             type, triplet) %>%
    summarise(count = n())
  
  d2.RT.epoch <- d2 %>%
    #filter(!str_detect(motor_triplet,'NA'), key_resp.corr == 1 ) %>%
    filter(!str_detect(triplet,'NA'),
           !str_detect(triplet,"(\\d)\\1{2}"),
           !str_detect(triplet,"(\\d).\\1"),
           ) %>%    
    group_by(epoch, frequency,  type) %>%
    summarise(RT = mean(key_resp.rt)*1000)  
  
  d2.RT.block <- d2 %>%
    #filter(!str_detect(motor_triplet,'NA'), key_resp.corr == 1 ) %>%
    filter(!str_detect(triplet,'NA'),
           !str_detect(triplet,"(\\d)\\1{2}"),
           !str_detect(triplet,"(\\d).\\1")
    ) %>%    
    group_by(block, frequency,  type) %>%
    summarise(RT = mean(key_resp.rt)*1000)  
  
  
  
    
  outcomes <- list(epochRT = d2.RT.epoch, 
                   frequency = d2.counts, 
                   raw = d2, 
                   blockRT = d2.RT.block )
  
  return(outcomes)
  
}

plotRT <- function(typeRT, typestr, data){
  
# data <- alloutcomes
#  typeRT <- 'epochRT'
#  typestr <- 'Epoch'
  xshift <- list(block = c(20,25), epoch = c(4,5)  )
  xticks <- list(block =30, epoch=6)
  xsep <- list(block = 5, epoch =1)
  
  fig <- ggplot(data$learn.out[[typeRT]], 
                          aes(x = .data[[typestr]], y = RT, 
                              color = frequency, 
                              shape = type)) +
    geom_point(aes(shape = type ), size = 3) +
    geom_line() +
    geom_point(data = data$motortest.out[[typeRT]], 
               aes(x = .data[[typestr]]+xshift[[typestr]][1], y = RT, 
                   color = frequency, 
                   shape = type), size=3) +
    geom_point(data = data$percepttest.out[[typeRT]], 
               aes(x = .data[[typestr]]+xshift[[typestr]][2], y = RT, 
                   color = frequency, 
                   shape = type), size=3) +
    labs(title = fn,
         x = typestr,
         y = "Reaction Time (ms)",
         color = "Frequency",
         linetype = "Type") +
    scale_x_continuous(breaks = seq(1, xticks[[typestr]], by = xsep[[typestr]]),  # Custom ticks every unit
                     labels = seq(1, xticks[[typestr]], by = xsep[[typestr]]))  # Custom labels corresponding to custom ticks
    theme_bw() +
    theme(text = element_text(size = 12), 
          plot.title = element_text(hjust = 0.5))  # Centering the title
  
    return(fig)  
  
}


droot <- '../data/'
allfn <- list.files(droot, pattern='*.csv')
fn <- allfn[3]

if (str_starts(fn, 'data')){
  sid <- str_extract(fn, "(?<=data_)[^.]+")
}else{
  sid <- str_extract(fn, "^[^_]+")
  
}


dpath <- file.path(droot, fn)

d <- import(dpath)

# Regular expression for the high frequency 
high_pattern_motor <- "2.3|3.1|1.0|0.2"
high_pattern_motor_rot <- "3.0|0.2|2.1|1.3"
high_pattern_percept <- "3.2|2.1|1.0|0.3"



d.learn <- d %>% filter(!is.na(learning_loop.thisN),
                        learning_loop.thisTrialN > 4)%>%
  select(which(!apply(is.na(d), 2, all))) %>%
  mutate(thisTrialN = learning_loop.thisTrialN)

d.motortest <- d %>% filter(!is.na(motor_testing_loop.thisN),
                        motor_testing_loop.thisTrialN > 4)%>%
  select(which(!apply(is.na(d), 2, all))) %>%
  mutate(thisTrialN = motor_testing_loop.thisTrialN)

d.percepttest <- d %>% filter(!is.na(percept_testing_loop.thisN),
                              percept_testing_loop.thisTrialN > 4)%>%
  select(which(!apply(is.na(d), 2, all))) %>%
  mutate(thisTrialN = percept_testing_loop.thisTrialN)


learn.out <- perform_summ(d.learn, high_pattern_motor)
motortest.out <- perform_summ(d.motortest, high_pattern_motor_rot )
percepttest.out <- perform_summ(d.percepttest, high_pattern_motor)

alloutcomes <- list(learn.out = learn.out, 
                    motortest.out = motortest.out, 
                    percepttest.out = percepttest.out)


f.learn.epoch <-  plotRT('epochRT', 'epoch', alloutcomes)
f.learn.block <-  plotRT('blockRT', 'block', alloutcomes)


show(f.learn.epoch)
ggsave(sprintf("figures/epoch/%s.jpg",sid), dpi =600)
show(f.learn.block)
ggsave(sprintf("figures/block/%s.jpg",sid), dpi = 600)

File: analysis/ana01.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from pathlib import Path
import sys

# Function to perform summary operations
def perform_summ(d, high_pattern):
    d = d.dropna(how='all', axis=1)  # Drop columns with all NAs
    d['triplet'] = d['correct_answer_index'].shift(2).astype(str) + d['correct_answer_index'].shift(1).astype(str) + d['correct_answer_index'].astype(str)
    d = d[['orientation_index', 'correct_answer_index', 'triplet', 'key_resp.corr', 'key_resp.rt', 'thisTrialN']]
    d['type'] = np.where(d['thisTrialN'] % 2 == 1, 'Pat', 'Ran')
    d['block'] =  (d.index // 80)
    d['epoch'] =  (d.index // 400)
    d['frequency'] = np.where(d['triplet'].str.contains(high_pattern), 'High', 'Low')

    d_counts = d[d['triplet'].str.contains('NA') == False].groupby(['epoch', 'frequency', 'type', 'triplet']).size().reset_index(name='count')
    
    d_rt_epoch = d[(d['triplet'].str.contains('NA') == False) & 
                   (d['triplet'].str.contains(r"(\d)\1{2}") == False) & 
                   (d['triplet'].str.contains(r"(\d).\1") == False) &
                   (d['key_resp.corr'] == 1)].groupby(['epoch', 'frequency', 'type']).agg(RT=('key_resp.rt', 'mean')).reset_index()
    d_rt_epoch['RT'] *= 1000

    d_rt_block = d[(d['triplet'].str.contains('NA') == False) & 
                   (d['triplet'].str.contains(r"(\d)\1{2}") == False) & 
                   (d['triplet'].str.contains(r"(\d).\1") == False) &
                   (d['key_resp.corr'] == 1)].groupby(['block', 'frequency', 'type']).agg(RT=('key_resp.rt', 'mean')).reset_index()
    d_rt_block['RT'] *= 1000

    return {'epochRT': d_rt_epoch, 'frequency': d_counts, 'raw': d, 'blockRT': d_rt_block}

# Function to plot RT data
def plot_rt(typeRT, typestr, data):
    fig, ax = plt.subplots(figsize=(10, 6))
    sns.lineplot(data=data['learn.out'][typeRT], x=typestr, y='RT', hue='frequency', style='type', markers=True, ax=ax)
    sns.scatterplot(data=data['motortest.out'][typeRT], x=typestr, y='RT', hue='frequency', style='type', markers=True, s=100, ax=ax)
    sns.scatterplot(data=data['percepttest.out'][typeRT], x=typestr, y='RT', hue='frequency', style='type', markers=True, s=100, ax=ax)
    ax.set_title(f'Reaction Time by {typestr}')
    ax.set_ylabel('Reaction Time (ms)')
    ax.set_xlabel(typestr)
    plt.legend(title='Frequency/Type')
    plt.grid(True)
    plt.show()

# Load data
droot = Path('../data/')
allfn = list(droot.glob('*.csv'))
fn = allfn[int(sys.argv[1])]
d = pd.read_csv(fn)

# Regular expression patterns
high_pattern_motor = "2.3|3.1|1.0|0.2"
high_pattern_percept = "3.2|2.1|1.0|0.3"

# Filtering and processing data
d_learn = d[d['learning_loop.thisN'].notna() & (d['learning_loop.thisTrialN'] > 4)]
d_learn['thisTrialN'] = d_learn['learning_loop.thisN']
d_motortest = d[d['motor_testing_loop.thisN'].notna() & (d['motor_testing_loop.thisTrialN'] > 4)]
d_motortest['thisTrialN'] = d_motortest['motor_testing_loop.thisN']
d_percepttest = d[d['percept_testing_loop.thisN'].notna() & (d['percept_testing_loop.thisTrialN'] > 4)]
d_percepttest['thisTrialN'] = d_percepttest['percept_testing_loop.thisTrialN']

# Perform summarization
learn_out = perform_summ(d_learn, high_pattern_motor)
motortest_out = perform_summ(d_motortest, high_pattern_motor)
percepttest_out = perform_summ(d_percepttest, high_pattern_percept)

all_outcomes = {'learn.out': learn_out, 'motortest.out': motortest_out, 'percepttest.out': percepttest_out}

# Plotting results
f_learn_epoch = plot_rt('epochRT', 'epoch', all_outcomes)
f_learn_block = plot_rt('blockRT', 'block', all_outcomes)
